dart/common/Uri.cpp

```diff
@@ -196,12 +206,125 @@ void Uri::clear()
 }
 
 //==============================================================================
+bool Uri::fromString(const std::string& _input)
+{
+  // This is regex is from Appendix B of RFC 3986.
+  static regex uriRegex(
+    R"END(^(([^:/?#]+):)?(//([^/?#]*))?([^?#]*)(\?([^#]*))?(#(.*))?)END",
+    regex::extended | regex::optimize);
+  static const size_t schemeIndex = 2;
+  static const size_t authorityIndex = 4;
+  static const size_t pathIndex = 5;
+  static const size_t queryIndex = 7;
+  static const size_t fragmentIndex = 9;
+
+  clear();
+
+  smatch matches;
+  if(!regex_match(_input, matches, uriRegex))
+    return false;
+
+  assert(matches.size() > schemeIndex);
+  const ssub_match& schemeMatch = matches[schemeIndex];
+  if(schemeMatch.matched)
+    mScheme = schemeMatch;
+
+  assert(matches.size() > authorityIndex);
+  const ssub_match& authorityMatch = matches[authorityIndex];
+  if(authorityMatch.matched)
+    mAuthority = authorityMatch;
+
+  assert(matches.size() > pathIndex);
+  const ssub_match& pathMatch = matches[pathIndex];
+  if(pathMatch.matched)
+  {
+    mPath = pathMatch;
+
+#ifdef _WIN32
+    if(mScheme && *mScheme == "file")
+    {
+      if(!mPath->empty() && mPath->at(0) == '/')
+        mPath = mPath->substr(1, mPath->size() - 1);
+    }
+#endif
+  }
+
+  assert(matches.size() > queryIndex);
+  const ssub_match& queryMatch = matches[queryIndex];
+  if(queryMatch.matched)
+    mQuery = queryMatch;
+
+  assert(matches.size() > fragmentIndex);
+  const ssub_match& fragmentMatch = matches[fragmentIndex];
+  if (fragmentMatch.matched)
+    mFragment = fragmentMatch;
+
+  return true;
+}
+
+//==============================================================================
+bool Uri::fromPath(const std::string& _path)
+{
+  // TODO(JS): We might want to check validity of _path.
+
+  static const std::string fileScheme("file://");
+
+  if (_path.compare(0, fileScheme.length(), fileScheme) != 0)
```

Comment:
> Is there a situation where recognizing and truncating the file:// scheme from the incoming string would break something?

I can't think of such a situation off the top of my head. My concern is that it makes the API more confusing. With two separate methods, the rule is simple: call `fromPath` if you have a path and `fromString` if you have a URI. With this change, you now have to worry about what scheme URI you have, too.

> I feel like the `fromPath` function should at least be smart enough to recognize when a string has a file:// scheme.

This is already handled by the `fromPathOrString` function which is, in my opinion, a more accurate name for what this function does. It also keeps these kinds of special cases contained to one function, so users that care about the difference between paths and URIs can bypass these types of checks.

In that case, I am comfortable modifying `fromStringOrPath` to use a heuristic to differentiate between absolute paths on Windows and general purpose URIs. That function could then serve as the "sane default" for users that don't care about the difference between paths and URIs.